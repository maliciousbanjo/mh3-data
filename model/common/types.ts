/** Status ailment types */
export enum Status {
  POISON = "poison",
  PARALYSIS = "paralysis",
  SLEEP = "sleep",
  STUN = "stun",
  EXHAUST = "exhaust",
}

/** Element types */
export type Element = "FIRE" | "WATER" | "THUNDER" | "ICE" | "DRAGON";

/** Levels of item/equipment rarity */
export type Rarity = 1 | 2 | 3 | 4 | 4 | 5 | 6 | 7;

/** Possible number of decoration slots on armors and weapons */
export type Slots = 0 | 1 | 2 | 3;