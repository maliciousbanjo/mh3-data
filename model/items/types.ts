/**
 * Ingame item eg; Mega Potion, Raw Meat, etc
 */
export interface Item {
  readonly id: number;
  readonly name: string;
}

export type ItemRecord = Record<Item["id"], Item>;

/**
 * Result of gathering/carving/quest reward
 *
 * Key:Value pair of an {@link Item} id and a percentage chance.
 */
export interface ItemResult {
  itemId: Item['id'];
  /** Percentage */
  chance: number;
}

/**
 * Quest reward item
 */
export interface QuestReward {
  itemId: Item['id'];
  /** Percentage */
  chance: number;
  /** Amount received */
  count: number;
}

/**
 * Amount needed of a particular item for crafting, trading, etc
 */
export interface ItemRequirement {
  itemId: Item['id'];
  amount: number;
}