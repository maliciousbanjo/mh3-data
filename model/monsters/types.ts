import { Status } from "../common";
import { QuestReward, ItemResult } from "../items";

export enum MonsterType {
  Herbvr = "herbivore",
  Nptron = "neopteron",
  Lynian = "lynian",
  Fish = "fish",
  PscWyv = "piscine wyvern",
  BrdWyv = "bird wyvern",
  FlyWyv = "flying wyvern",
  Levthn = "leviathan",
  BrtWyv = "brute wyvern",
  EldDrg = "elder dragon",
}

/**
 * Monster hitzone values for different damage types.
 * Used by {@link Hitzone} to match against a particular monster part
 */
export interface HitzoneValues {
  cut: number;
  impact: number;
  gun: number;
  fire: number;
  water: number;
  thunder: number;
  ice: number;
  dragon: number;
  stagger: number;
}

/**
 * Record of <Part: HitzoneValues>
 */
export type Hitzone = Record<string, HitzoneValues>;

/**
 * Collection of either a default hitzone or hitzones
 * based on a monster's status eg; enraged, flying, etc
 */
export interface HitzoneGroup {
  name: string;
  hitzones: Hitzone;
}

/**
 * Monster part that can be broken
 */
export interface Breakable {
  name: string;
  /** Low rank break rewards */
  low?: QuestReward[];
  /** High rank break rewards */
  high?: QuestReward[];
}

/**
 * Criteria and rewards for a monster being captured
 */
export interface CaptureInfo {
  /** Health percentage threshold for a monster to be capture-ready */
  health: number;
  /** Low rank capture rewards */
  low?: ItemResult[];
  /** High rank capture rewards */
  high?: ItemResult[];
}

/**
 * Status information for a particular {@link Status} ailment
 */
export interface StatusValues {
  /** Duration of status effect in seconds */
  duration: number;
  /** Resistance/buildup of status damage until effect activates */
  tolerance: {
    /** Value required for the first status effect */
    initial: number;
    /** Maximum resistance */
    max: number;
  };
  /** Amount of status damage recovered per interval */
  recovery: {
    /** Damage recovered */
    amount: number;
    /** Interval of recovery in seconds */
    wait: number;
  };
}

/**
 * Carvable monster part
 */
export interface CarveZone {
  name: string;
  /** Number of possible carves */
  count: number;
  /** Low rank carve */
  low?: ItemResult[];
  /** High rank carve */
  high?: ItemResult[];
}

/**
 * Extra unique data for a particular monster
 */
export interface Variant {
  name: string;
  carves?: CarveZone[];
  shiny?: ShinyDrop;
}

/**
 * Item result for a monster shiny drop
 */
export interface ShinyDrop {
  /** Action required to cause the shiny to drop */
  action: string;
  /** Low rank result */
  low?: ItemResult[];
  /** High rank result */
  high?: ItemResult[];
}

/**
 * Base properties for any monster
 */
export interface Monster {
  id: number;
  name: string;
  description: string;
  type: MonsterType;
  threat: number;
  /** 
   * List of hitzones. Will either contain a "default" hitzone group 
   * or multiple based on different monster states eg; flying, enraged, etc
   */
  hitzoneGroups: HitzoneGroup[];
  /** If undefined this monster should have a 'variant' property */
  shiny?: ShinyDrop;
  /** If undefined this monster should have a 'variant' property */
  carves?: CarveZone[];
  /** Optional support for varying shines or carves */
  variants?: Variant[];
}

/**
 * Properties for a large (boss) monster
 */
export interface LargeMonster
  extends Omit<Monster, "carves" | "type" | "variants"> {
  hp: number;
  type:
    | MonsterType.BrdWyv
    | MonsterType.FlyWyv
    | MonsterType.Levthn
    | MonsterType.BrtWyv
    | MonsterType.EldDrg;
  /** Quests that include this monster */
  quests: {
    // TODO: quest type. Will be closely tied to the quest data.
  };
  carves: CarveZone[];
  breakables: Breakable[];
  status?: {
    [Status.POISON]: StatusValues & {
      /** Damage taken by poison per interval */
      damage: {
        initial: number;
        total: number;
      };
    };
    [Status.PARALYSIS]: StatusValues;
    [Status.SLEEP]: StatusValues;
    [Status.STUN]: StatusValues;
    [Status.EXHAUST]?: Omit<StatusValues, "duration">;
  };
  /** {@link MonsterType.EldDrg} cannot be captured */
  capture?: CaptureInfo;
}

export type MonsterRecord = Record<Monster["id"], Monster>;
export type LargeMonsterRecord = Record<LargeMonster["id"], LargeMonster>;
