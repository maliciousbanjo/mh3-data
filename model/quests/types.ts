import { Item, QuestReward } from "../items";
import { Monster } from "../monsters";
import { MonsterLevel } from "../monster-levels";

export type QuestType =
  | "gather"
  | "hunt"
  | "capture"
  | "harvest"
  | "escort"
  | "multihunt"
  | "repel"
  | "slay";

export type QuestGoal =
  | "gathering"
  | "slay"
  | "capture"
  | "hunt"
  | "break"
  | "sever"
  | "sonic"
  | "repel"
  | "other"
  | "anthill";

export enum Location {
  DESERTED_ISLAND = "Deserted Island",
  SANDY_PLAINS = "Sandy Plains",
  FLOODED_FOREST = "Flooded Forest",
  TUNDRA = "Tundra",
  VOLCANO = "Volcano",
  UNDERWATER_RUIN = "Underwater Ruin",
  GREAT_DESERT = "Great Desert",
  SACRED_LAND = "Sacred Land",
  LAND_ARENA = "Land Arena",
  WATER_ARENA = "Water Arena",
}

/**
 * These properties were derived from the publicly-available MH3 Quest editor.
 * The values are incorrect/flawed. ZeSpyro
 */
interface BossInfo {
  type1: Monster["id"];
  /** Starting area? */
  type2: 0 | 1 | 2;
  /** TODO: Unidentified monster levels: 30, 55, 59, 60 */
  level: MonsterLevel | 30 | 55 | 59 | 60;

  size: number;
  /** controls random distribution of monster's HP */
  min: number;
  /** controls random distribution of monster's size */
  max: number;
}

/**
 * Base quest type
 *
 * It looks like different questTypes have different properties
 * 'target' points to differnt ids. Gather quests have IDs for items. Hunt/capture likely
 * have IDs for monsters, but the monsters themselves don't seem to have IDs.
 */
interface Quest {
  name: string;
  client: string;
  type: QuestType;
  goal: QuestGoal;
  /** Cost to initiate quest in zenny */
  contract: number;
  /** Maximum quest time in minutes */
  time: number;
  location: Location;
  /** If true, hunter will spawn in random location */
  randomspawn: boolean;
  /** Quest completion reward in zenny */
  reward: number;
  /** Amount of zenny taken from reward per faint */
  ko: number;
  /** Hunter Rank Points received for quest completion */
  hrp: number;
  secondaryRewards: QuestReward[];
  /** Key will be an ID for an item or monster depending on the {@link QuestType} */
  targets: Record<number, number>;
  /** Large monsters that can spawn in this quest */
  bosses: BossInfo[];
  // TODO: Will need to create different subquests types with different target properties
  subquests?: object;
}

type NumberRequired = number;

export interface GatheringQuest extends Quest {
  type: "gather";
  goal: "gathering";
  targets: Record<Item["id"], NumberRequired>;
}

export interface HarvestQuest extends Quest {
  type: "harvest";
  goal: "gathering";
  /** Harvest Ticket */
  targets: {
    396: 1;
  };
}

export interface HuntQuest extends Quest {
  type: "hunt";
  goal: "hunt" | "slay";
  targets: Record<Monster["id"], NumberRequired>;
  bosses: BossInfo[];
}

export interface MultiHuntQuest extends Quest {
  type: "multihunt";
  goal: "hunt";
  targets: Record<Monster["id"], NumberRequired>;
}

export interface CaptureQuest extends Quest {
  type: "capture";
  goal: "capture";
  targets: Record<Monster["id"], NumberRequired>;
  bosses: BossInfo[];
}

export interface EscortQuest extends Quest {
  type: "escort";
  goal: "slay";
  targets: Record<Monster["id"], NumberRequired>;
}

export interface ArenaQuest extends Omit<Quest, "secondaryRewards"> {
  type: "slay";
  goal: "hunt";
  targets: Record<Monster["id"], NumberRequired>;
  bosses: BossInfo[];
}

export interface RepelQuest extends Quest {
  type: "repel";
  goal: "repel" | "other";
  bosses: BossInfo[];
}

/** Quest star levels */
export enum StarLevel {
  OneStar = "1 Star",
  TwoStar = "2 Star",
  ThreeStar = "3 Star",
  FourStar = "4 Star",
  FiveStar = "5 Star",
  SixStar = "6 Star",
}

export interface QuestRegion {
  name: "Village" | "City";
  starLevels: {
    [StarLevel.OneStar]: Quest[];
    [StarLevel.TwoStar]: Quest[];
    [StarLevel.ThreeStar]: Quest[];
    [StarLevel.FourStar]: Quest[];
    [StarLevel.FiveStar]: Quest[];
    [StarLevel.SixStar]: Quest[];
  };
  arena: ArenaQuest[];
  events: Quest[];
}
