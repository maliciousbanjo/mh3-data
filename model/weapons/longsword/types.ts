import { Weapon, WeaponType } from "../types";

export type LongswordAttack =
  | "Unsheathe Attack"
  | "Step Stash"
  | "Downward Slash"
  | "Stab"
  | "Upward Slash"
  | "Sweep"
  | "Empty Spirit"
  | "Spirit 1"
  | "Spirit 2"
  | "Spirit 3"
  | "Spirit Finisher";

export type Longsword = Weapon<WeaponType.LONGSWORD>;
