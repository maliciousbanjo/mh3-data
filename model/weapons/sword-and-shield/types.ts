export type SwordAndShieldAttack =
  // land attacks
  | "Unsheathe Attack"
  | "Jumping Slash"
  | "Rising Slash"
  | "Regular Combo"
  | "Horizontal Combo"
  | "Shield Combo"
  | "Revolving Slice"
  | "Guard Slice"
  // water attacks
  | "Unsheathe Attack"
  | "Forward Slash"
  | "Normal Slash"
  | "Upward Slash"
  | "Uppercut Slash"
  | "Horizontal Slash"
  | "Shield Bash"
  | "Rising Slice"
  | "Revolving Slice"
  | "Guard Slice";
