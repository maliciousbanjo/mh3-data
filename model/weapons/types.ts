import { Element, Rarity, Slots, Status } from "../common";
import { ItemRequirement } from "../items";

export enum WeaponType {
  GREAT_SWORD = "Great Sword",
  SWORD_AND_SHIELD = "Sword and Shield",
  HAMMER = "Hammer",
  LANCE = "Lance",
  SWITCH_AXE = "Switch Axe",
  LONGSWORD = "Longsword",
}

/**
 * Properties specific to a hit with damage type 'cut'
*/
export interface CutHit {
  type: "cut";
  /** Multiplier used in damage calculations */
  power: number;}

/**
 * Properties specific to a hit with damage type 'impact'
*/
export interface ImpactHit {
  type: "impact";
  /** Multiplier used in damage calculations */
  power: number;
  /** Amount of KO damage dealt */
  ko: number;
}

export type Hit = CutHit | ImpactHit;

/**
 * Base type for an attack
 */
export interface Attack<T = string> {
  /** Human-readable attack name */
  name: T;
  hits: Hit[];
}

export interface AttackGroup<T> {
  /** Usage scenario @example default, underwater, axe mode */
  name: string;
  attacks: Attack<T>[];
}

/**
 * Collection of attacks and values used for determining wepaon damage
 */
export interface WeaponDamageProperties<T> {
  type: WeaponType;
  /** Used in the damage/item buff calculations */
  classModifier: number;
  attackGroups: AttackGroup<T>[];
}

/**
 * Properties of a weapon
 */
export interface Weapon<T extends WeaponType> {
  id: number;
  type: T;
  name: string;
  description: string;
  attack: number;
  element: Element | keyof typeof Status;
  elemAttack: number;
  /** If true requires armor skill 'Awaken' for element properties to take effect */
  awaken: boolean;
  /** Sharpness levels */
  sharpness: number[];
  /** Sharpness levels with armor skill 'Sharpness+1' */
  sharpnessUp: number[];
  affinity: number;
  slots: Slots;
  rarity: Rarity;
  /** In zenny */
  price: number;
  /** Optionally improves hunter defense */
  defense?: number;
  /** ID of previous weapon in path and what is required to upgrade it */
  upgradesFrom?: { weaponId: Weapon<WeaponType.GREAT_SWORD>['id'][]; materials: ItemRequirement[] };
  /** ID of next weapon in upgrade path */
  upgradesTo?: Weapon<WeaponType.GREAT_SWORD>['id'][];
  /** List of required items to create this weapon */
  create?: ItemRequirement[];
}