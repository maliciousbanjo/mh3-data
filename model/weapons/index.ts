// Data
import { GreatSwords, GreatSwordDamageProperties } from "./great-sword";
import {
  SwordAndShields,
  SwordAndShieldDamageProperties,
} from "./sword-and-shield";
import { Hammers, HammerDamageProperties } from "./hammer";
import { Lances, LanceDamageProperties} from './lance';

// Types
export type { GreatSwordTypes } from "./great-sword";
export type { SwordAndShieldTypes } from "./sword-and-shield";
export type { HammerTypes } from "./hammer";
export type { LanceTypes } from './lance';
export * from "./types";

export { GreatSwords, SwordAndShields, Hammers, Lances };

export const DamageProperties = {
  GreatSwordDamageProperties,
  SwordAndShieldDamageProperties,
  HammerDamageProperties,
  LanceDamageProperties
};
