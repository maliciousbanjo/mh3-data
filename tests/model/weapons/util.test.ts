import { WeaponType } from '../../../src/model/weapons/types';
import { getWeapon } from '../../../src/model/weapons/util';

describe('Weapon utils', () => {
  describe('getWeapon', () => {
    it('throws error when invalid ID provided', () => {
      expect(() => getWeapon(WeaponType.GREAT_SWORD, -1)).toThrow();
    });
    it('returns a valid great sword', () => {
      expect(getWeapon(WeaponType.GREAT_SWORD, 1)).toBeDefined();
    });
    it('returns a valid hammer', () => {
      expect(getWeapon(WeaponType.HAMMER, 1)).toBeDefined();
    });
    it('returns a valid lance', () => {
      expect(getWeapon(WeaponType.LANCE, 1)).toBeDefined();
    });
    it('returns a valid longsword', () => {
      expect(getWeapon(WeaponType.LONGSWORD, 1)).toBeDefined();
    });
    it('returns a valid switch axe', () => {
      expect(getWeapon(WeaponType.SWITCH_AXE, 1)).toBeDefined();
    });
    it('returns a valid sword and shield', () => {
      expect(getWeapon(WeaponType.SWORD_AND_SHIELD, 1)).toBeDefined();
    });
  });
});
