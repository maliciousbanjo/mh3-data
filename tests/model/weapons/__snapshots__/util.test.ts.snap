// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weapon utils calculateDamage Great Sword 1`] = `
[
  {
    "elementalDamage": 0.675,
    "koDamage": undefined,
    "rawDamage": 144.78750000000002,
    "totalDamage": 145,
  },
]
`;

exports[`Weapon utils calculateDamage Hammer 1`] = `
[
  {
    "elementalDamage": 0.675,
    "koDamage": 0,
    "rawDamage": 22.5,
    "totalDamage": 23,
  },
  {
    "elementalDamage": 0.675,
    "koDamage": 72,
    "rawDamage": 85.49999999999999,
    "totalDamage": 86,
  },
]
`;

exports[`Weapon utils calculateDamage Lance Cut hitzone 1`] = `
[
  {
    "elementalDamage": 2.025,
    "koDamage": undefined,
    "rawDamage": 25.515000000000004,
    "totalDamage": 27,
  },
  {
    "elementalDamage": 2.025,
    "koDamage": undefined,
    "rawDamage": 25.515000000000004,
    "totalDamage": 27,
  },
  {
    "elementalDamage": 2.025,
    "koDamage": undefined,
    "rawDamage": 27.337500000000002,
    "totalDamage": 29,
  },
]
`;

exports[`Weapon utils calculateDamage Lance Modified impact hitzone 1`] = `
[
  {
    "elementalDamage": 2.7,
    "koDamage": undefined,
    "rawDamage": 47.17440000000001,
    "totalDamage": 49,
  },
  {
    "elementalDamage": 2.7,
    "koDamage": undefined,
    "rawDamage": 47.17440000000001,
    "totalDamage": 49,
  },
  {
    "elementalDamage": 2.7,
    "koDamage": undefined,
    "rawDamage": 50.54400000000001,
    "totalDamage": 53,
  },
]
`;

exports[`Weapon utils calculateDamage Longsword Spirit Guage NONE 1`] = `
[
  {
    "elementalDamage": 2.115,
    "koDamage": undefined,
    "rawDamage": 25.485384375000002,
    "totalDamage": 27,
  },
]
`;

exports[`Weapon utils calculateDamage Longsword Spirit Guage RED 1`] = `
[
  {
    "elementalDamage": 2.115,
    "koDamage": undefined,
    "rawDamage": 33.1309996875,
    "totalDamage": 35,
  },
]
`;

exports[`Weapon utils calculateDamage Longsword Spirit Guage WHITE 1`] = `
[
  {
    "elementalDamage": 2.115,
    "koDamage": undefined,
    "rawDamage": 26.759653593750002,
    "totalDamage": 28,
  },
]
`;

exports[`Weapon utils calculateDamage Longsword Spirit Guage YELLOW 1`] = `
[
  {
    "elementalDamage": 2.115,
    "koDamage": undefined,
    "rawDamage": 28.0339228125,
    "totalDamage": 30,
  },
]
`;
